core/src/Cuda/Kokkos_Cuda_Instance.cpp

```diff
@@ -100,13 +100,21 @@ int cuda_kernel_arch() {
   int arch    = 0;
   int *d_arch = nullptr;
 
-  cudaMalloc(reinterpret_cast<void **>(&d_arch), sizeof(int));
-  cudaMemcpy(d_arch, &arch, sizeof(int), cudaMemcpyDefault);
+  CudaInternal::singleton().cuda_api_interface_safe_call<void **, size_t>(
+      &cudaMalloc, reinterpret_cast<void **>(&d_arch), sizeof(int));
+  CudaInternal::singleton()
+      .cuda_api_interface_safe_call<false, void *, const void *, size_t,
+                                    cudaMemcpyKind>(
+          &cudaMemcpy, d_arch, &arch, sizeof(int), cudaMemcpyDefault);
 
   query_cuda_kernel_arch<<<1, 1>>>(d_arch);
 
-  cudaMemcpy(&arch, d_arch, sizeof(int), cudaMemcpyDefault);
-  cudaFree(d_arch);
+  CudaInternal::singleton()
+      .cuda_api_interface_safe_call<false, void *, const void *, size_t,
+                                    cudaMemcpyKind>(
+          &cudaMemcpy, &arch, d_arch, sizeof(int), cudaMemcpyDefault);
+  Impl::CudaInternal::singleton().cuda_api_interface_safe_call<false, void *>(
+      &cudaFree, d_arch);
```

Comment:
Please pass the correct device id into this function (and use it) or add a FIXME_CUDA_MULTIPLE_DEVICES.

Reply:
I was passing `false` here since the first cuda api call in this function (which sets the device) ensures all the following calls are set correctly. I'll test again, but I think there should be no major performance gain by setting `false` the few times I do. I change it to all be the default (setting true).

There is only one place where we need `setCudaDevice=false` because we are in the process of querying for the device id. So we still need the option to set false. I'll tag it.